### PR TITLE
Created consensus.go_util.go  and edited consensuscore,rpc and types …

### PR DESCRIPTION
**Pull request by DarkLord017**
- I added some structs in types as in geth many fields were private and was different from alloy(from which helios was importing)
- in RPC Just edited a type for more correct usage
- In consensus_core made all public and implemented not implemented fields
- Wrote consensus/consensus.go and consensus/utils.go
- Previous PR issues have been resolved and are in this PR so please check 
- Also made a single array of bytes in utils for easy transformation from structs to byte array

### CODE REVIEW

**Comment by DarkLord017** [comment]
@star-gazer111 please review

**Comment by star-gazer111** [comment]
@DarkLord017 u were supposed to create a new utils.go in the utils folder that contains all the functions of the currently existing three different utils.go and remove the existing ones.

**Comment by DarkLord017** [comment]
I'm creating that one consensus me time lag gaya 
Pehle vo PR daalu fir ye pr daalu kya sir ?

**Comment by star-gazer111** [comment]
> I'm creating that one consensus me time lag gaya Pehle vo PR daalu fir ye pr daalu kya sir ? Sorry mujhe laga vo iske baad krna tha 🥲

What does this PR exactly does and which issue does it solves can u please reference?

**Comment by DarkLord017** [comment]
> > I'm creating that one consensus me time lag gaya Pehle vo PR daalu fir ye pr daalu kya sir ? Sorry mujhe laga vo iske baad krna tha 🥲
> 
> What does this PR exactly does and which issue does it solves can u please reference?

It solves #24 and also the changes of the previous PR #41

**Comment by star-gazer111** [comment]
> > > I'm creating that one consensus me time lag gaya Pehle vo PR daalu fir ye pr daalu kya sir ? Sorry mujhe laga vo iske baad krna tha 🥲
> > 
> > 
> > What does this PR exactly does and which issue does it solves can u please reference?
> 
> It solves #24 and also the changes of the previous PR #41

Okay thanks.
Also the one error in linter that's coming bcz of a function not being used in merkleproof.go. Where will this function which is unused for now be used in future?

**Comment by DarkLord017** [comment]
I'll check n tell u

**Comment by DarkLord017** [comment]
@star-gazer111 wait verify merkle proof vo use karke hi likh Deta

**Comment by DarkLord017** [comment]
now check @star-gazer111

**Comment by DarkLord017** [comment]
@star-gazer111 sir merkle tree library use karke kar liya utils me but vo function use mr nhi liya kyuki usme index ke bina hi validation ho rha

File: consensus/consensus.go [hunk 1]
```diff
@@ -4,48 +4,1065 @@ package consensus
...
+	chainID, err := in.RPC.ChainId()
+	if err != nil {
+		return err
+	}
+	if chainID != in.Config.Chain.ChainID {
+		return ErrIncorrectRpcNetwork
+	}
+	return nil
+}
+func (in *Inner) check_execution_payload(ctx context.Context, slot *uint64) (*consensus_core.ExecutionPayload, error) {
```
Comment (star-gazer111) [inline consensus/consensus.go:259]: and this function as `get_execution_payload`

File: consensus/consensus.go [hunk 2]
```diff
@@ -4,48 +4,1065 @@ package consensus
...
+	if err != nil {
+		return nil, err
+	}
+	endPayload := endBlock.Body.ExecutionPayload
+	prevParentHash := endPayload.ParentHash
+
+	// Create a wait group to manage concurrent fetching
+	var wg sync.WaitGroup
+	payloadsChan := make(chan interface{}, endSlot-startSlot+1)
+	errorChan := make(chan error, 1) // Buffer for one error
```
Comment (star-gazer111) [inline consensus/consensus.go:312]: wait, what is this?

File: consensus/consensus.go [hunk 3]
```diff
@@ -4,48 +4,1065 @@ package consensus
...
+}
+
+func (in *Inner) duration_until_next_update() time.Duration {
+	currentSlot := in.expected_current_slot()
+	nextSlot := currentSlot + 1
+	nextSlotTimestamp := nextSlot*12 + in.Config.Chain.GenesisTime
+
+	now := uint64(time.Now().Unix())
+	timeToNextSlot := int64(nextSlotTimestamp - now)
+	nextUpdate := timeToNextSlot + 4
```
Comment (star-gazer111) [inline consensus/consensus.go:496]: add comments regarding the 4 seconds time slot u are using.

File: consensus/consensus.go [hunk 4]
```diff
@@ -4,48 +4,1065 @@ package consensus
...
+	chainId := network.Chain.ChainID
+	var networkName config.Network
+	if chainId == 1 {
+		networkName = config.MAINNET
+	} else if chainId == 5 {
+		networkName = config.GOERLI
+	} else if chainId == 11155111 {
+		networkName = config.SEPOLIA
+	} else {
+		return errors.New("chain id not recognized")
```
[the remark this answers is not in the document]
Comment (DarkLord017) [reply consensus/consensus.go:218]: No taken from networks.go

File: consensus/consensus.go [hunk 5]
```diff
@@ -4,48 +4,1065 @@ package consensus
...
+	ErrInsufficientParticipation     = errors.New("insufficient participation")
+	ErrInvalidTimestamp              = errors.New("invalid timestamp")
+	ErrInvalidPeriod                 = errors.New("invalid period")
+	ErrNotRelevant                   = errors.New("update not relevant")
+	ErrInvalidFinalityProof          = errors.New("invalid finality proof")
+	ErrInvalidNextSyncCommitteeProof = errors.New("invalid next sync committee proof")
+	ErrInvalidSignature              = errors.New("invalid signature")
+)
+
+const MAX_REQUEST_LIGHT_CLIENT_UPDATES = 128
```
[the remark this answers is not in the document]
Comment (DarkLord017) [reply consensus/consensus.go:49]: Constants.go in Helios/consensus

File: consensus/consensus.go [hunk 6]
```diff
@@ -4,48 +4,1065 @@ package consensus
...
+	ErrInvalidUpdate                 = errors.New("invalid update")
+	ErrInsufficientParticipation     = errors.New("insufficient participation")
+	ErrInvalidTimestamp              = errors.New("invalid timestamp")
+	ErrInvalidPeriod                 = errors.New("invalid period")
+	ErrNotRelevant                   = errors.New("update not relevant")
+	ErrInvalidFinalityProof          = errors.New("invalid finality proof")
+	ErrInvalidNextSyncCommitteeProof = errors.New("invalid next sync committee proof")
+	ErrInvalidSignature              = errors.New("invalid signature")
+)
+
```
[the remark this answers is not in the document]
Comment (DarkLord017) [reply consensus/consensus.go:48]: https://github.com/a16z/helios/blob/master/consensus-core/src/errors.rs

File: consensus/consensus.go [hunk 7]
```diff
@@ -4,48 +4,1065 @@ package consensus
...
+	Config             *config.Config
+}
+type LightClientStore struct {
+	FinalizedHeader               consensus_core.Header
+	CurrentSyncCommitee           consensus_core.SyncCommittee
+	NextSyncCommitee              *consensus_core.SyncCommittee
+	OptimisticHeader              consensus_core.Header
+	PreviousMaxActiveParticipants uint64
+	CurrentMaxActiveParticipants  uint64
+}
```
[the remark this answers is not in the document]
Comment (DarkLord017) [reply consensus/consensus.go:85]: They were present in consensus_core module in helios but were not defined in consensus_core in selene so I defined  them here they were just commented out with not implemented

File: consensus/consensus.go [hunk 8]
```diff
@@ -4,48 +4,1065 @@ package consensus
...
+
+type GenericUpdate struct {
+	AttestedHeader          consensus_core.Header
+	SyncAggregate           consensus_core.SyncAggregate
+	SignatureSlot           uint64
+	NextSyncCommittee       *consensus_core.SyncCommittee
+	NextSyncCommitteeBranch *[]consensus_core.Bytes32
+	FinalizedHeader         consensus_core.Header
+	FinalityBranch          []consensus_core.Bytes32
+}
```
[the remark this answers is not in the document]
Comment (DarkLord017) [reply consensus/consensus.go:59]: They were present in consensus_core module in helios but were not defined in consensus_core in selene so I defined them here they were just commented out with not implemented

File: consensus/consensus.go (hunk 2 above)
Comment (DarkLord017) [reply consensus/consensus.go:312]: sync.waitgroup is join.all in rust to await for the completion for all 
payloadsChan for payloads and error chan for errors

Comment (DarkLord017) [reply consensus/consensus.go:304]: sync.Waitgroup in place for join.all in rust n payload chan for payloads and errors chan for error

Comment (star-gazer111) [reply consensus/consensus.go:304]: Cool

File: consensus/consensus.go [hunk 9]
```diff
@@ -220,9 +215,6 @@ func sync_all_fallback(inner *Inner, chainID uint64) error {
 
 	// Fetch the latest checkpoint from the network
 	checkpoint := checkpointFallback.FetchLatestCheckpoint(networkName)
-	if err != nil {
-		return err
-	}
```
Comment (star-gazer111) [inline consensus/consensus.go]: why are these lines removed?

Comment (DarkLord017) [reply consensus/consensus.go]: It was unnecessary already at line 193 it was defined
